app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsPagingSource.kt

```diff
@@ -0,0 +1,184 @@
+        // If the key is null this is straightforward, just return the most recent notifications.
+        val key = params.key
+            ?: return@coroutineScope mastodonApi.notifications(
+                limit = params.loadSize,
+                excludes = notificationFilter
+            )
+
+        // It's important to return *something* from this state. If an empty page is returned
+        // (even with next/prev links) Pager3 assumes there is no more data to load and stops.
+        //
+        // In addition, the Mastodon API does not let you fetch a page that contains a given key.
+        // You can fetch the page immediately before the key, or the page immediately after, but
+        // you can not fetch the page itself.
+
+        // First, try and get the notification itself, and the notifications immediately before
+        // it. This is so that a full page of results can be returned. Returning just the
+        // single notification means the displayed list can jump around a bit as more data is
+        // loaded.
+        //
+        // Make both requests, and wait for the first to complete.
+        val deferredNotification = async { mastodonApi.notification(id = key) }
+        val deferredNotificationPage = async {
+            mastodonApi.notifications(maxId = key, limit = params.loadSize, excludes = notificationFilter)
+        }
+
+        val notification = deferredNotification.await()
+        if (notification.isSuccessful) {
+            // If this was successful we must still check that the user is not filtering this type
+            // of notification, as fetching a single notification ignores filters. Returning this
+            // notification if the user is filtering the type is wrong.
+            notification.body()?.let { body ->
+                if (!notificationFilter.contains(body.type)) {
+                    // Notification is *not* filtered. We can return this, but need the next page of
+                    // notifications as well
+
+                    // Collect all notifications in to this list
+                    val notifications = mutableListOf(body)
+                    val notificationPage = deferredNotificationPage.await()
+                    if (notificationPage.isSuccessful) {
+                        notificationPage.body()?.let {
+                            notifications.addAll(it)
+                        }
+                    }
+
+                    // "notifications" now contains at least one notification we can return, and
+                    // hopefully a full page.
+
+                    // Build correct max_id and min_id links for the response. The "min_id" to use
+                    // when fetching the next page is the same as "key". The "max_id" is the ID of
+                    // the oldest notification in the list.
+                    val maxId = notifications.last().id
+                    val headers = Headers.Builder()
```

Comment:
A paging source deals directly with http stuff? This sounds like mixing very different levels?

Reply:
A `PagingSource` can deal directly with the network, as it does here.

Where there is also a database acting as the local source of truth a `RemoteMediator`  is also involved. That is not the case here because notifications are not currently maintained locally. They might be in the future, but the previous code didn't do it so this code doesn't do it either.

The Android docs that go in to this architecture in more detail start at https://developer.android.com/topic/libraries/architecture/paging/v3-overview